### ISSUE TITLE
[Flaky Test] Garden Controller Integration Test Suite

### ISSUE THREAD

**Issue by acumino** [gardener/gardener#7200]
**How to categorize this issue?**

/area testing
/kind flake

**Which test(s)/suite(s) are flaking**:
```
• [FAILED] [10.869 seconds]
Garden controller tests
/home/prow/go/src/github.com/gardener/gardener/test/integration/operator/garden/garden_test.go:44
  [It] should properly maintain the Reconciled condition
  /home/prow/go/src/github.com/gardener/gardener/test/integration/operator/garden/garden_test.go:100
```

**CI link**:
https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7035/pull-gardener-integration/1602607380611207168

**Reason for failure**:

**Anything else we need to know**:

**Comment by rfranzke** [comment]
Probably similar to https://github.com/gardener/gardener/issues/6933

**Comment by rfranzke** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7228/pull-gardener-integration/1604823985462185984

**Comment by rfranzke** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/batch/pull-gardener-integration/1605865569830047744

**Comment by rfranzke** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/batch/pull-gardener-integration/1605886955113943040

**Comment by rfranzke** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7293/pull-gardener-integration/1612368130741374976

**Comment by rfranzke** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/logs/ci-gardener-integration-release-v1-62/1612417668013363200

**Comment by shafeeqes** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7386/pull-gardener-integration/1617746807616442368

**Comment by acumino** [comment]
One of the reasons for the test to fail is https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7228/pull-gardener-integration/1604823985462185984#1:build-log.txt%3A283.

The above error occurs because sometimes the `etcd-main` and sometimes `etcd-events` deployment fails to get/create the secret `etcd-client`.
This is so because both processes try to achieve this thing at the same time (0.001-second difference). So one of them successfully creates the secret `etcd-client` but the other one fails to create and according to [this](https://github.com/gardener/gardener/blob/7dbfcc231ebdf0f36cd849c9bf7495d5a560b66d/pkg/utils/secrets/manager/generate.go#L123-L132) logic, it try to fetch it and fails cause secret is not yet sycned to cache.

One solution to prevent this issue can be to deploy `etcd-main` and `etcd-event` serialy instead of deploying it parallelly.

**Comment by rfranzke** [comment]
> One solution to prevent this issue can be to deploy `etcd-main` and `etcd-event` serialy instead of deploying it parallelly.

Would this really fix the problem consistently? There should not be much difference when comparing sequential/parallel deployment of these two components (it is quite fast, so I wouldn't expect a significant time difference when doing it sequentially).

Our default timeouts are `5s` for `Eventually` calls - this sounds like "large enough" for the caches to sync. @timebertt any hints/thoughts?

**Comment by acumino** [comment]
>Would this really fix the problem consistently? There should not be much difference when comparing sequential/parallel deployment of these two components (it is quite fast, so I wouldn't expect a significant time difference when doing it sequentially).

I ran more than 1000 runs and there were 0 flakes.

For cache sync to work, there should be another reconciliation that is not happening in our case. There is a logical issue with reconciliation logic, which I wanted to fix with the PR fixing this flake.
Here :> https://github.com/gardener/gardener/blob/7dbfcc231ebdf0f36cd849c9bf7495d5a560b66d/pkg/operator/controller/garden/reconciler.go#L107-L109
We are not respecting the original error and returning the error from `patchConditionToFalse` which can be nil even though the original reconciliation fails this should not be the case as it will not trigger the new reconciliation.

**Comment by timebertt** [comment]
> We are not respecting the original error and returning the error from `patchConditionToFalse` which can be nil even though the original reconciliation fails this should not be the case as it will not trigger the new reconciliation.

When looking at https://github.com/gardener/gardener/blob/3ba69b27aa358c45cb8f8fe3b054fc04d77ab089/pkg/operator/controller/garden/reconciler.go#L134-L139

we do return the original error from `r.reconcile` and thus go into exponential backoff, correct?

**Comment by rfranzke** [comment]
We looked into the flake together with @timebertt and found that there seem to be at least two different issues causing the flakes.

The [most recent flake](https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7386/pull-gardener-integration/1617746807616442368) is caused by too long interval/timeout settings when the `kubeapiserverexposure` component waits for readiness. We don't override this yet in the integration test setup, hence the default values (`5s` interval, `10m` timeout) were used - which is obviously way too long).

However, the other flakes are still to be investigated. @timebertt @acumino @ary1992 maybe we can look into this together next week?

**Comment by rfranzke** [comment]
Another one: https://prow.gardener.cloud/view/gs/gardener-prow/pr-logs/pull/gardener_gardener/7393/pull-gardener-integration/1618581634880114688

**Comment by timebertt** [comment]
We found the reason for the other flake:
The controller is performing retries of the garden reconciliation which re-creates resource-manager webhooks.
Because we only remove the webhooks once in the test, the second reconciliation fails deploying the etcds, because the patch to the `Etcd` objects runs into the webhook timeout.
Similar steps need to be faked in the test for every reconciliation, like removing the operation annotation from the `Etcd` objects.

We discussed the following mitigation:
- skip deploying resource-manager webhooks in integration tests
- fake external components' actions that need to be performed for every reconciliation in a fake reconciler (e.g., add a reconciler for removing the operation annotation)
  - things that only need to be faked once (e.g., patching the status) are kept in `Eventually`s like it is today
